### PR TITLE
feat: Add Support for qwen2.5 vl model inference(72b/32b/7b/3b) on Gpu And TPU

### PR DESCRIPTION
**Pull request by zkkython**
## Motivation
Add Support for qwen2.5 vl model inference(72b/32b/7b/3b) on Gpu And TPU
## Modifications

Based on current testing: Inference for the qwen2.5 vl model on both TPU and GPU is now supported, and GPU supports flash attention (non-native)

1 New Dependencies:
In addition to the original dependencies, the following dependencies have been added:
```
# Install CPU Version
pip install torch torchvision --index-url https://download.pytorch.org/whl/cpu

pip install flash-attn-jax --no-deps
pip install decord==0.6.0

# Install the python packages
pip install --upgrade pip setuptools packaging
pip install -e "python[gpu]"
```
2 Code Modifications Based on SGLang Logic

image

## Accuracy Tests

# 3 GPU Verification

### 3.1 Qwen2.5 VL 72B Model

```plain&#x20;text
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache python3 -u -m sgl_jax.launch_server \
  --model-path /data/hf/Qwen2.5-VL-72B-Instruct \
  --trust-remote-code \
  --dist-init-addr=0.0.0.0:10011 \
  --nnodes=1 \
  --tp-size=8 \
  --grammar-backend=none \
  --attention-backend=fa \
  --device=cuda \
  --random-seed=3 \
  --node-rank=0 \
  --mem-fraction-static=0.6 \
  --max-prefill-tokens=64 \
  --max-running-requests=16 \
  --download-dir=/tmp \
  --dtype=bfloat16 \
  --host 0.0.0.0 \
  --port 30000 \
  --enable-multimodal \
  --disable-overlap-schedule \
  --disable-precompile
 
```

#### 3.1.1 Text Inference:

```json
curl -X POST "http://localhost:30000/v1/chat/completions" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/data/hf/Qwen2.5-VL-72B-Instruct",
    "messages": [
      {
        "role": "user",
        "content": "hello"
      }
    ],
    "max_tokens": 10
  }'
```

Output:
image

#### 3.1.2 Image Understanding:

image

image

```json
curl -X POST "http://localhost:30000/v1/chat/completions" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/data/hf/Qwen2.5-VL-72B-Instruct",
    "messages": [
      {
        "role": "user",
        "content": [
          {"type": "text", "text": "Introduce the differences between these two images."}
          {
            "type": "image_url",
            "image_url": {
              "url": "1920-1200-2.png"
            }
          },
          {
            "type": "image_url",
            "image_url": {
              "url": "889-500.png"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

Output:

image

#### 3.1.3 Video Understanding

https://github.com/user-attachments/assets/f0bd1f25-a3ae-4ecc-bc59-b60c154ac771

```json
curl -X POST "http://localhost:30000/v1/chat/completions" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "/data/hf/Qwen2.5-VL-72B-Instruct",
    "messages": [
      {
        "role": "user",
        "content": [
          {"type": "text", "text": "Summarize the main content of this video"}
          {
            "type": "video_url",
            "video_url": {
              "url": "video.mp4"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

image

### 3.2 Qwen2.5 VL 32B Model

```plain&#x20;text
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache python3 -u -m sgl_jax.launch_server \
  --model-path /data/hf/Qwen2.5-VL-32B-Instruct \
  --trust-remote-code \
  --dist-init-addr=0.0.0.0:10011 \
  --nnodes=1 \
  --tp-size=8 \
  --grammar-backend=none \
  --attention-backend=fa \
  --device=cuda \
  --random-seed=3 \
  --node-rank=0 \
  --mem-fraction-static=0.6 \
  --max-prefill-tokens=64 \
  --max-running-requests=16 \
  --download-dir=/tmp \
  --dtype=bfloat16 \
  --host 0.0.0.0 \
  --port 30000 \
  --enable-multimodal \
  --disable-overlap-schedule \
  --disable-precompile
 
```

#### 3.2.1 Image Understanding

image

#### 3.2.2 Video Understanding

image

### 3.3 Qwen2.5 VL 7B Model

```plain&#x20;text
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache python3 -u -m sgl_jax.launch_server \
  --model-path /data/hf/Qwen2.5-VL-7B-Instruct \
  --trust-remote-code \
  --dist-init-addr=0.0.0.0:10011 \
  --nnodes=1 \
  --tp-size=1 \
  --grammar-backend=none \
  --attention-backend=fa \
  --device=cuda \
  --random-seed=3 \
  --node-rank=0 \
  --mem-fraction-static=0.6 \
  --max-prefill-tokens=1024 \
  --max-running-requests=16 \
  --download-dir=/tmp \
  --dtype=bfloat16 \
  --host 0.0.0.0 \
  --port 30000 \
  --enable-multimodal \
  --disable-overlap-schedule \
 
```

#### 3.3.1 Image Understanding

image
#### 3.3.2 Video Understanding

image

# 4 TPU Verification

Launch:

```plain&#x20;text
JAX_COMPILATION_CACHE_DIR=/tmp/jit_cache python3 -u -m sgl_jax.launch_server \
  --model-path Qwen/Qwen2.5-VL-3B-Instruct \
  --trust-remote-code \
  --dist-init-addr=0.0.0.0:10011 \
  --nnodes=1 \
  --tp-size=1 \
  --grammar-backend=none \
  --device=tpu \
  --random-seed=3 \
  --node-rank=0 \
  --mem-fraction-static=0.45 \
  --max-prefill-tokens=32 \
  --max-running-requests=1 \
  --download-dir=/tmp \
  --dtype=bfloat16 \
  --host 0.0.0.0 \
  --port 30000 \
  --enable-multimodal \
  --disable-overlap-schedule \
  --enable-single-process \
  --chunked-prefill-size=32 \
  --page-size=1
```

## 4.1 Inference Verification

```json
curl -X POST "http://localhost:30000/v1/chat/completions" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen2.5-VL-3B-Instruct",
    "messages": [
      {
        "role": "user",
        "content": [
          {"type": "text", "text": "Summarize the main content of this video"}
          {
            "type": "video_url",
            "video_url": {
              "url": "/home/gcpuser/sky_workdir/sglang-jax/video.mp4"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

image

```json
curl -X POST "http://localhost:30000/v1/chat/completions" \
  -H "Content-Type: application/json" \
  -d '{
    "model": "Qwen/Qwen2.5-VL-3B-Instruct",
    "messages": [
      {
        "role": "user",
        "content": [
          {"type": "text", "text": "Introduce the differences between these two images."}
          {
            "type": "image_url",
            "image_url": {
              "url": "/home/gcpuser/sky_workdir/sglang-jax/1920-1200-2.png"
            }
          },
          {
            "type": "image_url",
            "image_url": {
              "url": "/home/gcpuser/sky_workdir/sglang-jax/889-500.png"
            }
          }
        ]
      }
    ],
    "max_tokens": 300
  }'
```

image

# 5 SGlang vs Sglang-JAX Accuracy Comparison

From the evaluation results, the accuracy aligns with sglang
[sglang+sglang-jax_data_test.xlsx](https://github.com/user-attachments/files/24585737/sglang%2Bsglang-jax_data_test.xlsx)

## Checklist

- [ ] Please use English, otherwise it will be closed.
- [ ] The purpose of the PR, or link existing issues this PR will resolve.
- [ ] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py [hunk 1]
```diff
@@ -9,17 +9,22 @@
...
+from sgl_jax.srt.utils.jax_utils import device_array, is_tpu_runtime
+
+# Conditional imports based on runtime
+if is_tpu_runtime():
+    from sgl_jax.srt.kernels.ragged_paged_attention.ragged_paged_attention import (
+        ragged_paged_attention,
+    )
+    from sgl_jax.srt.speculative.eagle_util import EagleDraftInput
+else:
+    from flash_attn_jax import flash_mha, flash_mha_varlen
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/flashattention_backend.py:27]: Please add this to `pyproject`

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py [hunk 2]
```diff
@@ -532,6 +573,68 @@ def _ragged_paged_attention_with_fused_kv(*args):
...
+        v: jax.Array,
+        forward_batch: ForwardBatch,
+        token_to_kv_pool: KVCache,
+        kv_sharding: jax.NamedSharding,
+        layer_id: int,
+    ) -> tuple[jax.Array, jax.Array, jax.Array]:
+        """Get the kv cache from the forward batch (GPU path)."""
+        updated_layer = token_to_kv_pool.set_kv_buffer_legacy(
+            layer_id, forward_batch.out_cache_loc, k, v
+        )
```
Comment (JamesBrianD) [inline python/sgl_jax/srt/layers/attention/flashattention_backend.py:629]: Could you profile this to confirm that the update operation is efficient enough on GPU?

File: python/sgl_jax/srt/layers/attention/flashattention_backend.py [hunk 3]
```diff
@@ -95,11 +103,22 @@ def __init__(
...
 
     def get_forward_metadata(
         self,
         batch: ModelWorkerBatch,
     ):
+        # GPU backend doesn't need forward metadata
+        if not is_tpu_runtime():
+            return None
+
+        # NOTE: Removed undefined is_eagle, speculative_step_id, topk variables
```
Comment (pengchengneo) [inline python/sgl_jax/srt/layers/attention/flashattention_backend.py:116]: delete this eagle useless code

File: python/sgl_jax/srt/layers/multimodal.py
Comment (pengchengneo) [inline python/sgl_jax/srt/layers/multimodal.py:1]: this file name's scope is too large

**Comment by sii-xinglong** [comment]
Closing due to inactivity — this PR has been waiting for review for over 3 months. Please reopen if still relevant.
